src/backend/distributed/planner/merge_planner.c

```diff
@@ -1302,10 +1355,14 @@ ExtractMergeSourceRangeTableEntry(Query *query)
 	List *fromList = query->jointree->fromlist;
 
 	/* We should have only one RTE(MergeStmt->sourceRelation) in the from-list */
-	if (list_length(fromList) != 1)
+	if (list_length(fromList) != 1 && !missingOk)
 	{
 		ereport(ERROR, (errmsg("Unexpected source list in MERGE sql USING clause")));
 	}
+	else if (list_length(fromList) != 1)
```

Comment:
If we hit this, then it must be a bug. 

`list_length(fromList) must be 1 per Postgres design`

Even in this case

```
MERGE INTO target_cj t
USING (source_cj1 s1 INNER JOIN source_cj2 s2 ON sid1 = val2) s
ON t.tid = s.sid1 AND t.tid = 2
WHEN MATCHED THEN
        UPDATE SET src = src2
WHEN NOT MATCHED THEN
        DO NOTHING;
ERROR:  Source is not an explicit query
HINT:  Source query is a Join expression, try converting into a query as SELECT * FROM (..Join..)
```

It is "1"

Reply:
right, it seems I was being too conservative